### ISSUE TITLE
Review tt-xla wheel installation to check whether install step is accurate

### ISSUE THREAD

**Issue by einnesTT** [tenstorrent/tt-forge#458]
Thread listed below for reference - 

There was a question as to whether the docs are up-to-date for tt-xla. The decision was made previously to leave the 'final' instructions we want vs. changing them constantly or adding workarounds. This ticket is posted to show the work that was done on the question asked about the docs - 
- New cloud set-up 
- Update to Python 3.11 
- Try different wheel installs to share as an answer to the question 
- Find out when the docs will be back up-to-date - we are waiting on a commit that will make the instructions correct again and it hasn't been pushed just yet. 

Nick Smith
:spiral_calendar_pad:  [Yesterday at 2:33 PM](https://tenstorrent.slack.com/archives/C08H2NCQGHJ/p1758144813389779)
Hey all, are these docs up to date?  https://docs.tenstorrent.com/tt-xla/#installing-a-wheel-and-running-an-example
We have an external party that wants to run an llm e2e w/ pytorch and is also interested in dumping all of the mlir passes.  I'll do a bit more digging, just curious if anyone knows how to do this off hand.

7 replies

Grace Engelage
:spiral_calendar_pad:  [Yesterday at 2:55 PM](https://tenstorrent.slack.com/archives/C08H2NCQGHJ/p1758146126482939?thread_ts=1758144813.389779&cid=C08H2NCQGHJ)
Not sure that is is directly helpful but I have two torch single chip demos through tt-xla in a pr for tt-forge [here](https://github.com/tenstorrent/tt-forge/pull/453). When I tried the instructions in the docs it installed the previous release version of the wheel which doesn't have the support needed for torch models yet. On of the more recent versions of the wheel from [here](https://pypi.eng.aws.tenstorrent.com/pjrt-plugin-tt/) worked for me when testing locally (needed to upgrade the python version in the docker to python3.11).

Nick Smith
:spiral_calendar_pad:  [Yesterday at 3:45 PM](https://tenstorrent.slack.com/archives/C08H2NCQGHJ/p1758149123129359?thread_ts=1758144813.389779&cid=C08H2NCQGHJ)
Awesome thanks Grace!

Aleks Knezevic
:spiral_calendar_pad:  [Yesterday at 5:51 PM](https://tenstorrent.slack.com/archives/C08H2NCQGHJ/p1758156660637739?thread_ts=1758144813.389779&cid=C08H2NCQGHJ)
For dumping all of the graphs you can run with LOGGER_LEVEL=VERBOSE. For not quite as many graphs use DEBUG

Erikka Innes
  [Yesterday at 8:05 PM](https://tenstorrent.slack.com/archives/C08H2NCQGHJ/p1758164756249929?thread_ts=1758144813.389779&cid=C08H2NCQGHJ)
The docs show what the final outcome is supposed to be - there's a commit that needs to be made so that the command 'works' again. It hasn't been pushed through yet and I don't have more details about it than that.
To get it running until that command goes back to working, you would do this for a wheel:
pip install https://pypi.eng.aws.tenstorrent.com/pjrt-plugin-tt/pjrt_plugin_tt-0.4.0.dev20250917-cp311-cp311-linux_x86_64.whl
I tried one of the new demos from Grace for resnet, and it works as-is, however it had some things to note-
On a cloud VM without any other instructions to go with the demo, it uses the CPU not Tenstorrent. So you have to set that ahead of time, or in the code.
There's also some casting issues in it with tensors (if that matters) that slow it down a bit and it's hard to get all of them properly and cast them, I got some of them.
You also need to install FlagEmbedding and openmpi.
I thought the thing Grace added for other stuff for tt-torch demos that times how long everything takes was cool. I ran the posted version on the pull request she mentioned and I ran a version I made with the edits I listed above. The version I put updates in will run about 5-6 seconds faster (I pasted it below).
Everything also requires Python 3.11, which might be annoying for people to fix, so this information should be applied after you have Python 3.11 working.
I changed stuff in the resnet code I tried, if it's of interest to review does a few extra things... I have it so it tries to cast tensors to the right thing ahead of time (still misses some), so it directly uses Tenstorrent (you don't have to set it ahead of time), and I added the cool timing idea Grace put in other demos she created:
 # SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
#
# SPDX-License-Identifier: Apache-2.0

import os
import time
import warnings

# Set environment variable early to select Tenstorrent device
os.environ["PJRT_DEVICE"] = "TT"

import torch
from torch.utils._pytree import tree_map
import torch_xla         # Needed for torch_xla.device (we keep this import)
import torch_xla.core.xla_model as xm
import torch_xla.runtime as xr
from tt_torch.backend.backend import xla_backend
from FlagEmbedding import BGEM3FlagModel

def cast_int64_to_int32(obj):
    # Recursively cast any torch.Tensor of int64 dtype to int32 in nested struc>
    def cast_fn(x):
        if isinstance(x, torch.Tensor) and x.dtype == torch.int64:
            return x.to(torch.int32)
        return x
    return tree_map(cast_fn, obj)

def main():
    # Suppress specific FX/export warnings for cleaner output
    warnings.filterwarnings(
        "ignore",
        message="Attempted to insert a get_attr Node with no underlying referen>
    )
    warnings.filterwarnings(
        "ignore",
        message="Node lifted_tensor_0 target lifted_tensor_0.*"
    )

    # Explicitly set the XLA runtime device type to TT (Tenstorrent)
    xr.set_device_type("TT")

    # Load the model and set eval mode
    model = BGEM3FlagModel("BAAI/bge-m3").model
    model = model.eval()

    sentences = [
        "BGE M3 is an embedding model supporting dense retrieval, lexical match>
        "BM25 is a bag-of-words retrieval function that ranks a set of document>
    ]
    tokens = model.tokenizer(sentences, return_tensors="pt", padding=True, trun>

    # Cast any int64 tensors to int32 to avoid TT runtime warnings/errors
    tokens = cast_int64_to_int32(tokens)

    inputs = {
        "text_input": tokens,
        "return_dense": True,
        "return_sparse": True,
        "return_colbert_vecs": True,
    }

    # Compile the model with Tenstorrent XLA backend
    compiled_model = torch.compile(model, backend=xla_backend)

    def attempt_to_device(x):
        if hasattr(x, "to"):
            return x.to(device)
        return x

    # Correct device selection call here (use xm.xla_device)
    device = xm.xla_device(devkind="TT")
    compiled_model = compiled_model.to(device)
    inputs = tree_map(attempt_to_device, inputs)

    start_time = time.time()
    with torch.no_grad():
        outputs = compiled_model(**inputs)
    elapsed = time.time() - start_time

    print(f"Dense embeddings shape: {outputs['dense_vecs'].shape}")
    print(f"Sparse embeddings shape: {outputs['sparse_vecs'].shape}")
    print(f"ColBERT embeddings shape: {outputs['colbert_vecs'].shape}")
    print(f"Inference took {elapsed:.3f} seconds.")

if __name__ == "__main__":
    main()
I'm not sure if the samples go with accompanying documentation - some stuff can be removed or presented differently if you want to give people more info about why/what they're doing in the sample.

Erikka Innes
  [Yesterday at 8:19 PM](https://tenstorrent.slack.com/archives/C08H2NCQGHJ/p1758165571988639?thread_ts=1758144813.389779&cid=C08H2NCQGHJ)
NOTE ABOUT PYTHON:
You can use this command if you have Python 3.11 set up properly on your system:
pip install --pre pjrt-plugin-tt --extra-index-url https://pypi.eng.aws.tenstorrent.com/
It's just adding a flag to what's listed in the docs. (edited) 

Nick Smith
:spiral_calendar_pad:  [Today at 7:27 AM](https://tenstorrent.slack.com/archives/C08H2NCQGHJ/p1758205638698539?thread_ts=1758144813.389779&cid=C08H2NCQGHJ)
This is great, thank you so much Erikka!

Erikka Innes
  [Today at 7:33 AM](https://tenstorrent.slack.com/archives/C08H2NCQGHJ/p1758206039594219?thread_ts=1758144813.389779&cid=C08H2NCQGHJ)
Do we want to make users aware that they can get the very latest (that might not be stable) using the --pre flag? It sounds useful and it could prevent this issue in the future. If it doesn't count as a workaround only it's worth it imo.
There probably also needs to be a discussion about how we announce impactful changes like moving to Python 3.11 - I discovered it by having pre-commit break. For customer-developers, it would be a better experience if we're announcing stuff like this ahead of time so they know it's coming and don't just get a surprise. (I was out two weeks, so if we told people about it in an announcement, nm, I could have missed it.)
In addition to it changing how people use Forge, there's also using a cloud VM. If we gave a customer access to one, it's got Python 3.10 and they'd have to change it for this stuff.
